Enhance API to authorize users with system token authenticated requests.

abellotti opened the pull request:
This capability is needed for MiqLdap and External Authentication
Authentication Modes so users are authorized and auto-created
with system token authenticated requests.

- Added a User.admin?(userid) method
- Added a User.authorize_by_userid method
- Add support for authorize_user_by_userid for ldap, ldaps
- Add support for authorize_user_by_userid for httpd
- Add support to fetch user attrs for httpd

This solves the following BZ's:
https://bugzilla.redhat.com/show_bug.cgi?id=1400349
https://bugzilla.redhat.com/show_bug.cgi?id=1400350

abellotti commented:
Marking as WIP.
- more manual tests
- tire kicking with Central Admin
- rspecs

abellotti commented:
/cc @jvlcek @gtanzillo @carbonin @Fryguy 

Please review, Thanks!!

jvlcek commented on app/models/authenticator/httpd.rb:
```diff
@@ -4,15 +4,25 @@ def self.proper_name
+                      :lastname  => request.headers['X-REMOTE-USER-LASTNAME'],
+                      :email     => request.headers['X-REMOTE-USER-EMAIL']}
+        membership_list = (request.headers['X-REMOTE-USER-GROUPS'] || '').split(/[;:]/)
+      end
```
This looks like it could be DRY-ed up a little or perhaps create support methods. I suggest considering something like this:

```ruby
def get_ext_user_attrs(username)
    ext_user_attrs = get_user_attrs(username)
    user_attrs = {:username  => username,
                  :fullname  => ext_user_attrs["displayname"],
                  :firstname => ext_user_attrs["givenname"],
                  :lastname  => ext_user_attrs["sn"],
                  :email     => ext_user_attrs["mail"]}

    [user_attrs, MiqGroup.get_httpd_groups_by_user(username)]
end

def get_user_attrs_from_headers(username, request)
    user_attrs = {:username  => username,
                  :fullname  => request.headers['X-REMOTE-USER-FULLNAME'],
                  :firstname => request.headers['X-REMOTE-USER-FIRSTNAME'],
                  :lastname  => request.headers['X-REMOTE-USER-LASTNAME'],
                  :email     => request.headers['X-REMOTE-USER-EMAIL']}

    [user_attrs, (request.headers['X-REMOTE-USER-GROUPS'] || '').split(/[;:]/)]
end

def authorize_queue(username, request, options, *_args)
  if options[:authorize_only] == true
    user_attrs, membership_list = get_ext_user_attrs(username)
  else
    user_attrs, membership_list = get_user_attrs_from_headers(username, request)
  end 

  super(username, request, {}, user_attrs, membership_list)
end 

```

abellotti replied:
Not really DRY, but breaking it up in smaller methods is good, so 👍

jvlcek commented on app/models/authenticator.rb:
```diff
@@ -46,11 +60,11 @@ def authenticate(username, password, request = nil, options = {})
         username = normalize_username(username)
 
-        if _authenticate(username, password, request)
+        if options[:authorize_only] || _authenticate(username, password, request)
```
Cleverly way to avoid unnecessary modifications to this section of code. 👍

abellotti commented on app/models/authenticator.rb:
```diff
@@ -34,9 +34,23 @@ def uses_stored_password?
       false
     end
 
+    def authorize_user_by_userid?
```
naming is hard, was going by other method names in this file, i.e. authorize_queue?, no worry, I can change this.

abellotti commented:
@jvlcek I've updated the code as per your comments and added some specs. Did some manual tests here and still seems ok.   Could you try the PR changes in your environment ?  Thanks.

Fryguy commented on app/models/authenticator.rb:
```diff
@@ -34,9 +34,23 @@ def uses_stored_password?
+      options = {
+        :require_user   => true,
+        :authorize_only => true
+      }
```
Since you don't use these option elsewhere can you pass them directly to the next call.  It reads more explicit to me

```ruby
authenticate(userid, "", {},
  :require_user   => true,
  :authorize_only => true
)
```

Fryguy commented on app/models/authenticator.rb:
```diff
@@ -34,9 +34,23 @@ def uses_stored_password?
       false
     end
 
+    def can_authorize_user_by_userid?
```
Not sure I like the `can` prefix.  How about just `def authorize_user_by_userid?`?

jvlcek replied:
@Fryguy Funny because @abellotti had `autorize_user_by_userid?` and I thought that name was odd and suggested `can_authorize_user_by_userid?`.., Naming is hard. `can_user_be_authorized_by_userid?` <== Thoughts?

Fryguy commented on app/models/authenticator/httpd.rb:
```diff
@@ -48,5 +52,45 @@ def update_user_attributes(user, _username, identity)
       user.email      = user_attrs[:email] unless user_attrs[:email].blank?
     end
+
+    def get_user_details_by_userid(username)
```
Unless this is being written to conform to an existing pattern, can you remove teh `get` prefix, as it's generally avoided in Ruby code. e.g. here it could just be `user_details_by_userid`

jvlcek replied:
If the `get_` is removed from `get_user_details_by_userid` is should also be removed from: 
`get_user_details_from_headers`

Fryguy commented:
Something feels clunky about this, but it might just be the existing way that authenticator is designed.  Can we review in person?

jvlcek commented:
FYI: I've successfully tested this change with MiQLdap and External Auth.

abellotti commented:
With latest fix to group matching, tested successfully with Central Admin, doing a VM power op and user auto-created in lower region.

Fryguy commented:
Had a live review with @abellotti and I am :+1: with this PR pending the method renames.  We came up with 

`authorize_user_by_userid` -> `authorize_user`
`can_authorize_user_by_userid?` -> `user_authorizable_without_authentication?`

abellotti commented:
PR updated, @jvlcek and I retesting (just in case), if all goes well, I'll unwip.  Thanks @Fryguy

jvlcek commented:
FYI: I've successfully **_`re`-_**tested this change with MiQLdap and External Auth.
LGTM! 👍  💯  Good Job @abellotti

chessbyte commented:
Euwe Backport details:
```sh
$ git log -1
commit 766e56776d716f151ee23ccec9c3c46d3cdc3c37
Author: Gregg Tanzillo <gtanzill@redhat.com>
Date:   Mon Dec 12 16:58:10 2016 -0500

    Merge pull request #13068 from abellotti/api_sys_auth_authorize_user
    
    Enhance API to authorize users with system token authenticated requests.
    (cherry picked from commit fadf640a190c8d6dc97a84e3143163a07e5ef2c6)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1400349
    https://bugzilla.redhat.com/show_bug.cgi?id=1400350
```
